OpenRA.Mods.Cnc/Traits/ConyardChronoReturn.cs

```diff
@@ -0,0 +1,251 @@
+		[Desc("Facing of the returned actor.")]
+		public readonly int Facing = 96;
+
+		public readonly string ChronoshiftSound = "chrono2.aud";
+
+		[Desc("The color the bar of the 'return-to-origin' logic has.")]
+		public readonly Color TimeBarColor = Color.White;
+
+		public object Create(ActorInitializer init) { return new ConyardChronoReturn(init, this); }
+	}
+
+	public class ConyardChronoReturn : INotifyCreated, ITick, ISync, ISelectionBar, IDeathActorInitModifier,
+		ITransformActorInitModifier, INotifyBuildComplete, INotifySold, INotifyTransform
+	{
+		readonly ConyardChronoReturnInfo info;
+		readonly WithSpriteBody wsb;
+		readonly Health health;
+		readonly Actor self;
+		readonly string faction;
+
+		IPositionable positionable;
+		ConditionManager conditionManager;
+		int conditionToken = ConditionManager.InvalidConditionToken;
+
+		Actor chronosphere;
+		int duration;
+		bool buildComplete;
+		bool selling;
+
+		[Sync]
+		int returnTicks = 0;
+
+		[Sync]
+		CPos origin;
+
+		[Sync]
+		bool triggered;
+
+		public ConyardChronoReturn(ActorInitializer init, ConyardChronoReturnInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+
+			health = self.Trait<Health>();
+			wsb = self.TraitsImplementing<WithSpriteBody>().Single(w => w.Info.Name == info.Body);
+			faction = init.Contains<FactionInit>() ? init.Get<FactionInit, string>() : self.Owner.Faction.InternalName;
+
+			if (init.Contains<ChronoshiftReturnInit>())
+				returnTicks = init.Get<ChronoshiftReturnInit, int>();
+
+			if (init.Contains<ChronoshiftDurationInit>())
+				duration = init.Get<ChronoshiftDurationInit, int>();
+
+			if (init.Contains<ChronoshiftOriginInit>())
+				origin = init.Get<ChronoshiftOriginInit, CPos>();
+
+			if (init.Contains<ChronoshiftChronosphereInit>())
+				chronosphere = init.Get<ChronoshiftChronosphereInit, Actor>();
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+			positionable = self.TraitOrDefault<IPositionable>();
+		}
+
+		void TriggerVortex()
+		{
+			if (!string.IsNullOrEmpty(info.Condition))
+				conditionToken = conditionManager.GrantCondition(self, info.Condition);
+
+			triggered = true;
+
+			// Don't override the selling animation
+			if (selling)
+				return;
+
+			wsb.PlayCustomAnimation(self, info.Sequence, () =>
+			{
+				triggered = false;
+				if (conditionToken != ConditionManager.InvalidConditionToken)
+					conditionToken = conditionManager.RevokeCondition(self, conditionToken);
+			});
+		}
+
+		CPos? ChooseBestDestinationCell(CPos destination)
+		{
+			if (chronosphere == null)
+				return null;
+
+			if (positionable.CanEnterCell(destination) && chronosphere.Owner.Shroud.IsExplored(destination))
+				return destination;
+
+			var max = chronosphere.World.Map.Grid.MaximumTileSearchRange;
+			foreach (var tile in self.World.Map.FindTilesInCircle(destination, max))
+				if (chronosphere.Owner.Shroud.IsExplored(tile) && positionable.CanEnterCell(tile))
```

Comment:
Why are we checking `IsExplored`? (I.e. is there a technical reason, as this doesn't make sense to me otherwise.)

Reply:
This is copied from the `Teleport` activity.  I agree that most of that code doesn't make sense, but changing that is out of scope of this PR.